internal/healthcheck/healthcheck_test.go

```diff
@@ -0,0 +1,117 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package healthcheck
+
+import (
+	"context"
+	"net/http"
+	"testing"
+
+	"github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/proxy"
+)
+
+const testPort = "8080"
+
+// Test to verify that when the proxy client is up, the liveness endpoint writes 200.
+func TestLiveness(t *testing.T) {
+	proxyClient := &proxy.Client{}
+	hc := NewHealthCheck(proxyClient, testPort)
+	defer hc.Close(context.Background()) // Close health check upon exiting the test.
+
+	resp, err := http.Get("http://localhost:" + hc.port + livenessPath)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if resp.StatusCode != 200 {
+		t.Errorf("got status code %v instead of 200", resp.StatusCode)
+	}
+}
+
+// Test to verify that when startup has not finished, the readiness endpoint writes 500.
+func TestStartupFail(t *testing.T) {
+	proxyClient := &proxy.Client{}
+	hc := NewHealthCheck(proxyClient, testPort)
+	defer hc.Close(context.Background())
+
+	resp, err := http.Get("http://localhost:" + hc.port + readinessPath)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if resp.StatusCode != 500 {
+		t.Errorf("got status code %v instead of 500", resp.StatusCode)
+	}
+}
+
+// Test to verify that when startup has finished, and MaxConnections has not been reached,
+// the readiness endpoint writes 200.
+func TestStartupPass(t *testing.T) {
+	proxyClient := &proxy.Client{}
+	hc := NewHealthCheck(proxyClient, testPort)
+	defer hc.Close(context.Background())
+
+	// Simulate the proxy client completing startup.
+	hc.NotifyReadyForConnections()
+
+	resp, err := http.Get("http://localhost:" + hc.port + readinessPath)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if resp.StatusCode != 200 {
+		t.Errorf("got status code %v instead of 200", resp.StatusCode)
+	}
+}
+
+// Test to verify that when startup has finished, but MaxConnections has been reached,
+// the readiness endpoint writes 500.
+func TestMaxConnectionsReached(t *testing.T) {
+	proxyClient := &proxy.Client{
+		MaxConnections: 10,
+	}
+	hc := NewHealthCheck(proxyClient, testPort)
+	defer hc.Close(context.Background())
+
+	hc.NotifyReadyForConnections()
+	proxyClient.ConnectionsCounter = proxyClient.MaxConnections // Simulate reaching the limit for maximum number of connections
+
+	resp, err := http.Get("http://localhost:" + hc.port + readinessPath)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if resp.StatusCode != 500 {
+		t.Errorf("got status code %v instead of 500", resp.StatusCode)
+	}
+}
+
+// Test to verify that after closing a healthcheck, its liveness endpoint serves
+// an error.
+func TestCloseHealthCheck(t *testing.T) {
+	proxyClient := &proxy.Client{}
+	hc := NewHealthCheck(proxyClient, testPort)
+
+	resp, err := http.Get("http://localhost:" + hc.port + livenessPath)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if resp.StatusCode != 200 {
+		t.Errorf("got status code %v instead of 200", resp.StatusCode)
+	}
+
+	hc.Close(context.Background())
```

Comment:
Close should be deferred before the t.Fatal (in general, try to defer the Close as soon as logically possible)